# Import Dependencies

In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', 500)

# Read in Data

In [2]:
# long_CS= pd.read_csv("resources\SEDA_geodist_long_CS_v21.csv")

FileNotFoundError: File b'resources\\SEDA_geodist_long_CS_v21.csv' does not exist

In [ ]:
# long_gcs= pd.read_csv("resources\SEDA_geodist_long_GCS_v21.csv")

In [5]:
long_naep= pd.read_csv("SEDA_geodist_long_NAEP_v21.csv")

In [ ]:
# long_state= pd.read_csv("resources\SEDA_geodist_long_State_v21.csv")

In [ ]:
# pool_cs= pd.read_csv("resources\SEDA_geodist_pool_CS_v21.csv")

In [ ]:
# poolsub_cs= pd.read_csv("resources\SEDA_geodist_poolsub_CS_v21.csv")

# Columns in the data

In [ ]:
print(len(long_CS.columns))
print(len(long_gcs.columns))
print(len(long_naep.columns))
print(len(long_state.columns))
print(len(pool_cs.columns))
print(len(poolsub_cs.columns))

In [ ]:
columns_df= pd.DataFrame({
    "long_CS": pd.Series(long_CS.columns),
    "long_gcs": pd.Series(long_gcs.columns),
    "long_naep": pd.Series(long_naep.columns),
    "long_state": pd.Series(long_state.columns),
    "pool_cs": pd.Series(pool_cs.columns),
    "poolsub_cs": pd.Series(poolsub_cs.columns)
})

In [ ]:
columns_df

# Off with their heads!

In [ ]:
long_CS.head()

In [ ]:
long_gcs.head()

In [6]:
long_naep.head()

leaidC                 leaname  fips stateabb  grade  year subject  \
0  100002  Alabama Youth Services     1       AL      8  2009     ela   
1  100002  Alabama Youth Services     1       AL      8  2009    math   
2  100002  Alabama Youth Services     1       AL      8  2011     ela   
3  100002  Alabama Youth Services     1       AL      8  2011    math   
4  100002  Alabama Youth Services     1       AL      8  2012     ela   

   totgyb_all     mn_all     sd_all  mn_all_se  sd_all_se  totgyb_asn  mn_asn  \
0        29.0  209.40759  32.880596   7.390182   0.585985         NaN     NaN   
1        36.0  217.34308  33.107059   7.648670   0.802523         NaN     NaN   
2        29.0  230.81302  31.958628   7.168319   0.721690         NaN     NaN   
3        27.0  221.81458  31.811636   8.290004   0.677796         NaN     NaN   
4        33.0  227.06160  32.230938   6.740346   0.519832         NaN     NaN   

   sd_asn  mn_asn_se  sd_asn_se  totgyb_blk     mn_blk     sd_blk  mn_blk_se  \
0     NaN        NaN        NaN         NaN        NaN        NaN        NaN   
1     NaN        NaN        NaN        20.0  208.98753  30.545776  10.567958   
2     NaN        NaN        NaN         NaN        NaN        NaN        NaN   
3     NaN        NaN        NaN         NaN        NaN        NaN        NaN   
4     NaN        NaN        NaN         NaN        NaN        NaN        NaN   

   sd_blk_se  totgyb_fem  mn_fem  sd_fem  mn_fem_se  sd_fem_se  totgyb_hsp  \
0        NaN         NaN     NaN     NaN        NaN        NaN         NaN   
1   0.816991         NaN     NaN     NaN        NaN        NaN         NaN   
2        NaN         NaN     NaN     NaN        NaN        NaN         NaN   
3        NaN         NaN     NaN     NaN        NaN        NaN         NaN   
4        NaN         NaN     NaN     NaN        NaN        NaN         NaN   

   mn_hsp  sd_hsp  mn_hsp_se  sd_hsp_se  totgyb_mal     mn_mal     sd_mal  \
0     NaN     NaN        NaN        NaN        26.0  207.21269  33.474255   
1     NaN     NaN        NaN        NaN        33.0  215.39435  35.041161   
2     NaN     NaN        NaN        NaN        27.0  229.81973  32.197830   
3     NaN     NaN        NaN        NaN        26.0  223.95609  33.710880   
4     NaN     NaN        NaN        NaN        33.0  224.74576  32.058556   

   mn_mal_se  sd_mal_se  totgyb_wht  mn_wht  sd_wht  mn_wht_se  sd_wht_se  \
0   7.928550   0.650314         NaN     NaN     NaN        NaN        NaN   
1   8.297847   0.899029         NaN     NaN     NaN        NaN        NaN   
2   7.477520   0.773098         NaN     NaN     NaN        NaN        NaN   
3   8.844554   0.766845         NaN     NaN     NaN        NaN        NaN   
4   6.718442   0.573944         NaN     NaN     NaN        NaN        NaN   

   mn_wag  mn_wag_se  mn_wbg  mn_wbg_se  mn_whg  mn_whg_se  mn_mfg  mn_mfg_se  
0     NaN        NaN     NaN        NaN     NaN        NaN     NaN        NaN  
1     NaN        NaN     NaN        NaN     NaN        NaN     NaN        NaN  
2     NaN        NaN     NaN        NaN     NaN        NaN     NaN        NaN  
3     NaN        NaN     NaN        NaN     NaN        NaN     NaN        NaN  
4     NaN        NaN     NaN        NaN     NaN        NaN     NaN        NaN

In [ ]:
long_state.head()

In [ ]:
pool_cs.head()

In [ ]:
poolsub_cs.head()

# Describe the datasets

In [8]:
# desc_long_CS= long_CS.describe()
# desc_long_gcs= long_gcs.describe()
desc_long_naep= long_naep.describe()
# desc_long_state= long_state.describe()
# desc_pool_cs= pool_cs.describe()
# desc_poolsub_cs= poolsub_cs.describe()

In [ ]:
desc_long_CS

In [ ]:
desc_long_gcs

In [ ]:
desc_long_naep

In [ ]:
desc_long_state

In [ ]:
desc_pool_cs

In [ ]:
desc_poolsub_cs

# Trying to make sense of the data

## NAEP

In [9]:
desc_long_naep

leaidC           fips          grade           year  \
count  8.258160e+05  825816.000000  825816.000000  825816.000000   
mean   3.003463e+06      29.903256       5.439089    2011.844344   
std    1.457768e+06      14.573536       1.694571       1.959203   
min    1.000020e+05       1.000000       3.000000    2009.000000   
25%    1.809690e+06      18.000000       4.000000    2010.000000   
50%    2.930660e+06      29.000000       5.000000    2012.000000   
75%    4.106300e+06      41.000000       7.000000    2013.000000   
max    5.606240e+06      56.000000       8.000000    2015.000000   

          totgyb_all         mn_all         sd_all      mn_all_se  \
count  825814.000000  825814.000000  825814.000000  825814.000000   
mean      328.138101     245.368960      30.441846       3.896056   
std      1166.280140      24.817281       5.184859       1.943400   
min        20.000000     114.207150       1.565124       0.363533   
25%        55.000000     228.196378      26.897011       2.418443   
50%       119.000000     245.202560      30.546568       3.439700   
75%       277.000000     262.570257      33.708389       5.099976   
max     76800.000000     375.458440     147.766400      61.499256   

           sd_all_se    totgyb_asn        mn_asn        sd_asn     mn_asn_se  \
count  825814.000000  85523.000000  85523.000000  85523.000000  85523.000000   
mean        2.408928    122.689382    264.213928     34.154915      6.580929   
std         1.575808    483.938421     30.265167      6.139830      2.941361   
min         0.248818     20.000000     99.107666      6.260830      0.529085   
25%         1.045357     29.000000    243.636095     30.374425      4.555845   
50%         2.302719     47.000000    263.100400     34.027737      6.535615   
75%         3.374068     97.000000    283.998410     37.234781      8.154646   
max        44.424931  12214.000000    441.867650    148.027940     74.957466   

          sd_asn_se     totgyb_blk         mn_blk         sd_blk  \
count  85523.000000  183171.000000  183171.000000  183171.000000   
mean       2.920028     224.051460     228.758834      29.498467   
std        2.652629     719.344149      22.675072       4.759979   
min        0.388104      20.000000     129.180240       3.078496   
25%        1.009118      36.000000     213.230470      26.277265   
50%        1.852622      72.000000     229.219020      29.588207   
75%        4.200429     170.000000     244.799475      32.436369   
max       69.068367   22678.000000     314.946560      90.725815   

           mn_blk_se      sd_blk_se     totgyb_fem         mn_fem  \
count  183171.000000  183171.000000  684574.000000  684574.000000   
mean        4.592179       2.268489     188.532721     248.236700   
std         2.103669       1.726871     621.838991      24.501244   
min         0.513911       0.312495      20.000000     123.957090   
25%         2.881899       0.750063      39.000000     230.850810   
50%         4.330511       1.956334      74.000000     247.870575   
75%         6.143965       3.353889     159.000000     265.586045   
max        41.399170      37.110298   37799.000000     403.577090   

              sd_fem      mn_fem_se      sd_fem_se     totgyb_hsp  \
count  684574.000000  684574.000000  684574.000000  222124.000000   
mean       29.727471       4.546764       2.357524     252.636262   
std         4.981161       2.008899       1.750996    1028.612933   
min         3.804324       0.488723       0.262531      20.000000   
25%        26.303948       2.984426       0.658222      33.000000   
50%        29.795918       4.273438       2.253879      66.000000   
75%        32.748504       5.934157       3.520308     177.000000   
max       103.046170      59.726406      42.617218   38791.000000   

              mn_hsp         sd_hsp      mn_hsp_se      sd_hsp_se  \
count  222124.000000  222124.000000  222124.000000  222124.000000   
mean      233.040101      29.822799       4.694741       2.091

The NAEP score is derived from the [National Assessment of Educational Progress](https://nces.ed.gov/nationsreportcard/ "NAEP Nations Report Card"). The test is on a 0-500 scale, which we can see reflected in the means of all the schools *(mn_all)*.

In [ ]:
print(f'Minimum AVERAGE score for all schools: {desc_long_naep.loc["min","mn_all"]}')
print(f'Maximum AVERAGE score for all schools: {desc_long_naep.loc["max","mn_all"]}')

The above shows that the AVERAGE maximum score for any school is 375.5. This isn't to say that the highest score that a student received on the test is that number. Rather, that school could've had someone score the full 500 and another student who scored in the double digits. This is important to keep in mind while parsing through the data.

The other measures use various statistical techniques (HETOP and HOMOP, for example) to derive their normalized values, which I did not want to dig much deeper into. I believe using the NAEP score should be sufficient for the purposes of this data analysis.

Another interesting story comes from the "gaps" (disparity in average scores between sub-groups within a year/class/subject)

In [ ]:
# Columns which contain values for the gaps between scores of:
# White v Asian students
#       v Black students
#       v Hispanic students
# Male v Female students

gap_cols= ["mn_wag","mn_wbg","mn_whg","mn_mfg"]

desc_long_naep.loc[:,gap_cols]

Looking at the min and max from the *mn_wag* column, it looks like there is a vast disparity between the scores of White students and Asian students (min: -131.1, max: 137.2). In fact, this suggests the opposite.

These scores are derived by subtracting the first subgroup ("W") from the second ("A"). So it appears that in at least one school, Asian student scores are __on average higher__ than their White peers by 131 points. The converse is true given the max value as well.

Armed with this information, we can see that there does indeed appear to be a disparity between White and Black students' as well as White and Hispanic students' scores. The same holds for Male v Female scores, with the disparity favoring Female students.

In [12]:
abbrev_naep = long_naep[['leaidC', 'leaname', 'fips', 'stateabb', 'grade', 'year', 'subject',
       'totgyb_all', 'mn_all', 'totgyb_asn', 'mn_asn', 'totgyb_blk', 'mn_blk',
        'totgyb_fem', 'mn_fem','totgyb_hsp','totgyb_mal', 'mn_mal','totgyb_wht', 'mn_wht', 'mn_wag',
                                        'mn_wbg', 'mn_whg', 'mn_mfg',]]
abbrev_naep.head()

leaidC                 leaname  fips stateabb  grade  year subject  \
0  100002  Alabama Youth Services     1       AL      8  2009     ela   
1  100002  Alabama Youth Services     1       AL      8  2009    math   
2  100002  Alabama Youth Services     1       AL      8  2011     ela   
3  100002  Alabama Youth Services     1       AL      8  2011    math   
4  100002  Alabama Youth Services     1       AL      8  2012     ela   

   totgyb_all     mn_all  totgyb_asn  mn_asn  totgyb_blk     mn_blk  \
0        29.0  209.40759         NaN     NaN         NaN        NaN   
1        36.0  217.34308         NaN     NaN        20.0  208.98753   
2        29.0  230.81302         NaN     NaN         NaN        NaN   
3        27.0  221.81458         NaN     NaN         NaN        NaN   
4        33.0  227.06160         NaN     NaN         NaN        NaN   

   totgyb_fem  mn_fem  totgyb_hsp  totgyb_mal     mn_mal  totgyb_wht  mn_wht  \
0         NaN     NaN         NaN        26.0  207.21269         NaN     NaN   
1         NaN     NaN         NaN        33.0  215.39435         NaN     NaN   
2         NaN     NaN         NaN        27.0  229.81973         NaN     NaN   
3         NaN     NaN         NaN        26.0  223.95609         NaN     NaN   
4         NaN     NaN         NaN        33.0  224.74576         NaN     NaN   

   mn_wag  mn_wbg  mn_whg  mn_mfg  
0     NaN     NaN     NaN     NaN  
1     NaN     NaN     NaN     NaN  
2     NaN     NaN     NaN     NaN  
3     NaN     NaN     NaN     NaN  
4     NaN     NaN     NaN     NaN

In [13]:
naep_ela = abbrev_naep.loc[abbrev_naep.subject == "ela"]
naep_ela.head()

leaidC                 leaname  fips stateabb  grade  year subject  \
0  100002  Alabama Youth Services     1       AL      8  2009     ela   
2  100002  Alabama Youth Services     1       AL      8  2011     ela   
4  100002  Alabama Youth Services     1       AL      8  2012     ela   
6  100005        Albertville City     1       AL      3  2009     ela   
8  100005        Albertville City     1       AL      3  2010     ela   

   totgyb_all     mn_all  totgyb_asn  mn_asn  totgyb_blk  mn_blk  totgyb_fem  \
0        29.0  209.40759         NaN     NaN         NaN     NaN         NaN   
2        29.0  230.81302         NaN     NaN         NaN     NaN         NaN   
4        33.0  227.06160         NaN     NaN         NaN     NaN         NaN   
6       310.0  204.40581         NaN     NaN         NaN     NaN       161.0   
8       317.0  206.01045         NaN     NaN         NaN     NaN         NaN   

      mn_fem  totgyb_hsp  totgyb_mal     mn_mal  totgyb_wht     mn_wht  \
0        NaN         NaN        26.0  207.21269         NaN        NaN   
2        NaN         NaN        27.0  229.81973         NaN        NaN   
4        NaN         NaN        33.0  224.74576         NaN        NaN   
6  206.97821       124.0       149.0  201.54659       178.0  211.53754   
8        NaN       117.0         NaN        NaN       193.0  211.12666   

   mn_wag  mn_wbg     mn_whg    mn_mfg  
0     NaN     NaN        NaN       NaN  
2     NaN     NaN        NaN       NaN  
4     NaN     NaN        NaN       NaN  
6     NaN     NaN  16.336487 -5.431625  
8     NaN     NaN  14.736786       NaN

In [14]:
naep_math = abbrev_naep.loc[abbrev_naep.subject == "math"]
naep_math.head()

leaidC                 leaname  fips stateabb  grade  year subject  \
1  100002  Alabama Youth Services     1       AL      8  2009    math   
3  100002  Alabama Youth Services     1       AL      8  2011    math   
5  100002  Alabama Youth Services     1       AL      8  2012    math   
7  100005        Albertville City     1       AL      3  2009    math   
9  100005        Albertville City     1       AL      3  2010    math   

   totgyb_all     mn_all  totgyb_asn  mn_asn  totgyb_blk     mn_blk  \
1        36.0  217.34308         NaN     NaN        20.0  208.98753   
3        27.0  221.81458         NaN     NaN         NaN        NaN   
5        31.0  229.63074         NaN     NaN         NaN        NaN   
7       310.0  216.63632         NaN     NaN         NaN        NaN   
9       317.0  228.81721         NaN     NaN         NaN        NaN   

   totgyb_fem     mn_fem  totgyb_hsp  totgyb_mal     mn_mal  totgyb_wht  \
1         NaN        NaN         NaN        33.0  215.39435         NaN   
3         NaN        NaN         NaN        26.0  223.95609         NaN   
5         NaN        NaN         NaN        31.0  229.88904         NaN   
7       160.0  216.87672       125.0       150.0  215.30283       177.0   
9         NaN        NaN       117.0         NaN        NaN       192.0   

      mn_wht  mn_wag  mn_wbg    mn_whg    mn_mfg  
1        NaN     NaN     NaN       NaN       NaN  
3        NaN     NaN     NaN       NaN       NaN  
5        NaN     NaN     NaN       NaN       NaN  
7  220.40680     NaN     NaN  8.775574 -1.573898  
9  233.30249     NaN     NaN  9.994629       NaN

In [15]:
naep_ela_no_na = naep_ela.dropna(how="any")
naep_ela_no_na.head()

leaidC      leaname  fips stateabb  grade  year subject  totgyb_all  \
172  100007  Hoover City     1       AL      3  2009     ela       986.0   
176  100007  Hoover City     1       AL      3  2011     ela       974.0   
178  100007  Hoover City     1       AL      3  2012     ela       983.0   
180  100007  Hoover City     1       AL      3  2013     ela       995.0   
182  100007  Hoover City     1       AL      3  2014     ela       997.0   

        mn_all  totgyb_asn     mn_asn  totgyb_blk     mn_blk  totgyb_fem  \
172  214.54395        80.0  220.27069       214.0  190.93071       457.0   
176  217.66377        64.0  222.10507       221.0  193.69196       485.0   
178  215.21539        81.0  220.94724       215.0  191.52635       477.0   
180  213.73804        72.0  220.89682       234.0  196.55606       492.0   
182  219.03290        71.0  229.80255       233.0  197.46252       447.0   

        mn_fem  totgyb_hsp  totgyb_mal     mn_mal  totgyb_wht     mn_wht  \
172  217.31458        50.0       529.0  212.36569       642.0  224.37381   
176  219.95566        66.0       489.0  214.78072       619.0  231.69139   
178  219.43379        62.0       506.0  211.59489       619.0  226.93578   
180  215.47179        75.0       503.0  211.50256       577.0  224.43622   
182  222.84422        56.0       550.0  215.92966       598.0  228.89838   

       mn_wag     mn_wbg     mn_whg    mn_mfg  
172  4.103119  33.443100  29.331573 -4.948883  
176  9.586319  37.999435  42.134506 -5.174942  
178  5.988541  35.409424  41.469589 -7.838898  
180  3.539398  27.880157  40.126190 -3.969223  
182 -0.904175  31.435852  38.209534 -6.914566

In [17]:
naep_ela_no_na.count()

leaidC        27516
leaname       27516
fips          27516
stateabb      27516
grade         27516
year          27516
subject       27516
totgyb_all    27516
mn_all        27516
totgyb_asn    27516
mn_asn        27516
totgyb_blk    27516
mn_blk        27516
totgyb_fem    27516
mn_fem        27516
totgyb_hsp    27516
totgyb_mal    27516
mn_mal        27516
totgyb_wht    27516
mn_wht        27516
mn_wag        27516
mn_wbg        27516
mn_whg        27516
mn_mfg        27516
dtype: int64

In [20]:
len(naep_ela_no_na.index)

27516

In [16]:
naep_math_no_na = naep_math.dropna(how="any")
naep_math_no_na.head()

leaidC      leaname  fips stateabb  grade  year subject  totgyb_all  \
173  100007  Hoover City     1       AL      3  2009    math       987.0   
177  100007  Hoover City     1       AL      3  2011    math       983.0   
179  100007  Hoover City     1       AL      3  2012    math       985.0   
181  100007  Hoover City     1       AL      3  2013    math       995.0   
183  100007  Hoover City     1       AL      3  2014    math      1000.0   

        mn_all  totgyb_asn     mn_asn  totgyb_blk     mn_blk  totgyb_fem  \
173  222.00957        80.0  228.68329       213.0  205.59877       459.0   
177  223.60043        67.0  231.37659       222.0  207.22340       488.0   
179  223.30418        81.0  234.07726       215.0  205.46187       478.0   
181  226.40253        72.0  245.31400       234.0  213.29327       492.0   
183  230.91217        71.0  244.89435       232.0  216.53716       448.0   

        mn_fem  totgyb_hsp  totgyb_mal     mn_mal  totgyb_wht     mn_wht  \
173  220.91368        54.0       528.0  222.98395       640.0  228.69203   
177  222.21732        71.0       495.0  225.15512       619.0  231.22810   
179  223.13036        63.0       507.0  223.36121       620.0  229.63614   
181  225.05655        75.0       503.0  227.76344       577.0  231.85593   
183  230.94514        56.0       552.0  230.68436       601.0  236.18112   

        mn_wag     mn_wbg     mn_whg    mn_mfg  
173   0.008743  23.093262  27.785645  2.070267  
177  -0.148483  24.004700  22.464676  2.937805  
179  -4.441116  24.174271  28.265060  0.230850  
181 -13.458069  18.562653  20.885620  2.706894  
183  -8.713226  19.643967  17.980286 -0.260788

In [21]:
len(naep_math_no_na.index)

26423